File: app/src/main/java/com/friendlypos/reimprimirRecibos/fragment/ReimprimirReciboResumenFragment.java
package com.friendlypos.reimprimirRecibos.fragment;

import android.content.DialogInterface;
import android.os.Bundle;
import androidx.appcompat.app.AlertDialog;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.Recibos.modelo.receipts;
import com.friendlypos.Recibos.modelo.recibos;
import com.friendlypos.app.broadcastreceiver.BluetoothStateChangeReceiver;
import com.friendlypos.application.util.Functions;
import com.friendlypos.application.util.PrinterFunctions;
import com.friendlypos.distribucion.fragment.BaseFragment;
import com.friendlypos.principal.modelo.Clientes;
import com.friendlypos.reimprimirRecibos.activity.ReimprimirRecibosActivity;

import org.sufficientlysecure.htmltextview.HtmlTextView;

import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import io.realm.Realm;
import io.realm.RealmResults;

public class ReimprimirReciboResumenFragment extends BaseFragment {

    public HtmlTextView text;
    receipts recibo_actualizado;
    @Bind(R.id.btnReimprimirReciboNuevo)
    public ImageButton btnReimprimirFactura;
    BluetoothStateChangeReceiver bluetoothStateChangeReceiver;
    ReimprimirRecibosActivity activity;
    String facturaId = "";
    int slecTAB;
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bluetoothStateChangeReceiver = new BluetoothStateChangeReceiver();
        bluetoothStateChangeReceiver.setBluetoothStateChangeReceiver(getContext());
        activity = new ReimprimirRecibosActivity();
    }
    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getActivity().unregisterReceiver(bluetoothStateChangeReceiver);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_reimprimir_recibo_resumen, container, false);
        text = (HtmlTextView) rootView.findViewById(R.id.html_textReciboNuevo);
        ButterKnife.bind(this, rootView);


        btnReimprimirFactura.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {

                if(bluetoothStateChangeReceiver.isBluetoothAvailable()== true) {

                    LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
                    View promptView = layoutInflater.inflate(R.layout.prompt_imprimir_recibos, null);

                    final AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
                    alertDialogBuilder.setView(promptView);
                    final CheckBox checkbox = (CheckBox) promptView.findViewById(R.id.checkbox);

                    final TextView label = (TextView) promptView.findViewById(R.id.promtClabelRecibosImp);
                    label.setText("Escriba el número de impresiones requeridas");

                    final EditText input = (EditText) promptView.findViewById(R.id.promtCtextRecibosImp);

                    alertDialogBuilder.setCancelable(false);
                    alertDialogBuilder.setPositiveButton("OK", new DialogInterface.OnClickListener() {

                        public void onClick(DialogInterface dialog, int id) {

                            String cantidadImpresiones = input.getText().toString();

                            PrinterFunctions.imprimirReimpRecibosTotal(recibo_actualizado, getActivity(), 1, cantidadImpresiones);
                            Toast.makeText(getActivity(), "Reimprimir recibo", Toast.LENGTH_SHORT).show();
                        }
                    });
                    alertDialogBuilder.setNegativeButton("Cancel",
                            new DialogInterface.OnClickListener() {

                                public void onClick(DialogInterface dialog, int id) {
                                    dialog.cancel();
                                }
                            });

                    AlertDialog alertD = alertDialogBuilder.create();
                    alertD.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
                    alertD.show();

                }
                else if(bluetoothStateChangeReceiver.isBluetoothAvailable() == false){
                    Functions.CreateMessage(getActivity(), "Error", "La conexión del bluetooth ha fallado, favor revisar o conectar el dispositivo");
                }
            }
        });

        return rootView;
    }


    private void getHtmlPreview() {
        try {

            Realm realm = Realm.getDefaultInstance();
            Clientes clientes = realm.where(Clientes.class).equalTo("id", recibo_actualizado.getCustomer_id()).findFirst();

            String nombreCliente = clientes.getFantasyName();
            Log.d("nombreCliente", nombreCliente);
            String preview = "";


            if (recibo_actualizado != null) {


                preview += "<h5>" + "Recibos" + "</h5>";

                preview += "<a><b>A nombre de:</b> " + nombreCliente + "</a><br><br>";
             /*   preview += "<a><b>" + padRight("# Referencia", 25) + padRight("# Factura", 25)+ "</b></a><br>";
                preview += "<a><b>" + padRight("Monto total", 25) + padRight("Monto Pagado", 25)+ "</b></a><br>";
                preview += "<a><b>" + padRight("Total en abonos", 35) + "</b></a><br>";
                preview += "<a><b>" + padRight("Total restante", 35) + "</b></a><br>";*/
               /* preview += "<a><b>" + padRight("Monto total", 35)+ "</b></a><br>";
                preview += "<a><b>" + padRight("Monto Pagado", 10) + padRight("Monto restante", 10) + "</b></a><br>";*/

                preview += "<a>------------------------------------------------<a><br>";

                preview += getPrintDistTotal(recibo_actualizado.getCustomer_id());

            }
            else {
                preview += "<center><h2>Seleccione la factura a ver</h2></center>";
            }
            text.setHtmlFromString(preview, new HtmlTextView.LocalImageGetter());
        }
        catch (Exception e) {
            String preview = "<center><h2>Seleccione la factura a ver cath</h2></center>";
            text.setHtmlFromString(preview, new HtmlTextView.LocalImageGetter());
            Log.d("adsdad", e.getMessage());
        }
    }

    public static String padRight(String s, double n) {
        String centeredString;
        double pad = (n + 4) - s.length();

        if (pad > 0) {
            String pd = Functions.paddigTabs((int) (pad / 2.0));
            centeredString = "\t" + s + "\t" + pd;
            System.out.println("pad: " + "|" + centeredString + "|");
        }
        else {
            centeredString = "\t" + s + "\t";
        }
        return centeredString;
    }

    private String getPrintDistTotal(String idVenta) {
        String send = "";

        Realm realm1 = Realm.getDefaultInstance();
        RealmResults<receipts> result = realm1.where(receipts.class).equalTo("customer_id", idVenta).equalTo("reference",recibo_actualizado.getReference()).findAll();
        Log.d("recibosresult", result+"");
        if (result.isEmpty()) {
            send = "No hay recibos emitidos";
        }
        else {

                List<receipts> salesList1 = realm1.where(receipts.class).equalTo("customer_id", idVenta)
                        .equalTo("reference",recibo_actualizado.getReference()).findAll();

                Log.d("getReference", salesList1.get(0).getReference());

               recibos recibos = realm1.where(recibos.class).equalTo("numeration", salesList1.get(0).getNumeration()).findFirst();

                Log.d("getNumeration", recibos.getNumeration() +"");

                String numeroReferenciaReceipts = salesList1.get(0).getReference();
                String numeracionReceipts = salesList1.get(0).getNumeration();
                double pagadoReceipts =  salesList1.get(0).getMontoCanceladoPorFactura();
                String pagadoSReceipts = String.format("%,.2f", pagadoReceipts);

                double total = recibos.getTotal();
                String totalS = String.format("%,.2f", total);

                double restante = salesList1.get(0).getPorPagarReceipts();
                    String restanteS = String.format("%,.2f", restante);

                double totalAbonos = salesList1.get(0).getBalance();
                String totalAbonosS = String.format("%,.2f", totalAbonos);

               send += /*"# Referencia" + padRight(numeroReferenciaReceipts, 20) + "# Factura" + padRight(numeracionReceipts, 20)+ "<br>" +
                       "Monto total" + padRight(totalS, 40) + "Monto Pagado" + padRight(pagadoSReceipts, 40) + "<br>" +*/

                       "<a><b>" +  padRight("# Referencia:", 20 )+"</b></a>"+ padRight(numeroReferenciaReceipts, 20) + "<br>" +
                       "<a><b>" + padRight("# Factura:", 30 ) +"</b></a>"+ padRight(numeracionReceipts, 20) + "<br>" +
                       "<a><b>" + padRight("Monto Pagado:", 20 ) +"</b></a>" + padRight(pagadoSReceipts, 20) + "<br>" +
                       "<a><b>" + padRight("Monto total:", 30 ) +"</b></a>" + padRight(totalS, 20) + "<br>" +
                       "<a><b>" + padRight("Total en abonos:", 20 ) + "</b></a>" +padRight(totalAbonosS, 20) + "<br>" +
                        "<a><b>" + padRight("Total restante:", 25 ) +"</b></a>" + padRight(restanteS, 20)+"<br>";

               send += "<a>------------------------------------------------<a><br>";

                Log.d("FACTPRODTODFAC", send + "");


            realm1.close();
        }
        return send;
    }

    public void updateData() {
        slecTAB = ((ReimprimirRecibosActivity) getActivity()).getSelecReciboTab();
        if (slecTAB == 1) {

            facturaId =  ((ReimprimirRecibosActivity) getActivity()).getInvoiceIdReimprimirRecibo();


            final Realm realm3 = Realm.getDefaultInstance();
            realm3.executeTransaction(new Realm.Transaction() {

                @Override
                public void execute(Realm realm3) {

                    recibo_actualizado = realm3.where(receipts.class).equalTo("reference", facturaId).findFirst();

                    realm3.close();
                }
            });


            getHtmlPreview();
            Log.d("FACTURAIDTOTALIZAR", recibo_actualizado.getCustomer_id()+"");
            Log.d("FACTURAIDTOTALIZAR", facturaId);
        }
        else {
            Log.d("nadaTotalizarupdate",  "nadaTotalizarupdate");
        }

    }
}

File: app/src/main/java/com/friendlypos/reimprimirRecibos/fragment/ReimprimirReciboResumenFragment.kt
package com.friendlysystemgroup.friendlypos.reimprimirRecibos.fragment

import android.content.DialogInterface
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.friendlysystemgroup.friendlypos.R
import com.friendlysystemgroup.friendlypos.Recibos.modelo.receipts
import com.friendlysystemgroup.friendlypos.Recibos.modelo.recibos
import com.friendlysystemgroup.friendlypos.app.broadcastreceiver.BluetoothStateChangeReceiver
import com.friendlysystemgroup.friendlypos.application.util.Functions
import com.friendlysystemgroup.friendlypos.application.util.PrinterFunctions
import com.friendlysystemgroup.friendlypos.databinding.FragmentReimprimirReciboResumenBinding
import com.friendlysystemgroup.friendlypos.distribucion.fragment.BaseFragment
import com.friendlysystemgroup.friendlypos.principal.modelo.Clientes
import com.friendlysystemgroup.friendlypos.ReimprimirRecibos.activity.ReimprimirRecibosActivity
import io.realm.Realm
import io.realm.RealmResults
import com.friendlypos.util.HtmlTextView
import com.friendlypos.util.LocalImageGetter

class ReimprimirReciboResumenFragment : BaseFragment() {
    private var binding: FragmentReimprimirReciboResumenBinding? = null
    private var bluetoothStateChangeReceiver: BluetoothStateChangeReceiver? = null
    private var activity: ReimprimirRecibosActivity? = null
    private var reciboActualizado: receipts? = null
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bluetoothStateChangeReceiver = BluetoothStateChangeReceiver().apply {
            setBluetoothStateChangeReceiver(requireContext())
        }
        activity = getActivity() as? ReimprimirRecibosActivity
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            bluetoothStateChangeReceiver?.let {
                requireActivity().unregisterReceiver(it)
            }
        } catch (e: Exception) {
            Log.e("ReciboResumen", "Error al desregistrar receptor", e)
        }
        bluetoothStateChangeReceiver = null
        activity = null
        binding = null
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentReimprimirReciboResumenBinding.inflate(inflater, container, false)
        return binding?.root
    }
    
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupButtons()
    }
    
    private fun setupButtons() {
        binding?.btnReimprimirReciboNuevo?.setOnClickListener {
            if (bluetoothStateChangeReceiver?.isBluetoothAvailable == true) {
                showPrintDialog()
            } else {
                Functions.CreateMessage(
                    requireActivity(),
                    "Error",
                    "La conexión del bluetooth ha fallado, favor revisar o conectar el dispositivo"
                )
            }
        }
    }
    
    private fun showPrintDialog() {
        val layoutInflater = LayoutInflater.from(requireActivity())
        val promptView = layoutInflater.inflate(R.layout.prompt_imprimir_recibos, null)
        
        val alertDialogBuilder = AlertDialog.Builder(requireActivity()).apply {
            setView(promptView)
            setCancelable(false)
            setPositiveButton("OK") { _, _ ->
                val input = promptView.findViewById<EditText>(R.id.promtCtextRecibosImp)
                val cantidadImpresiones = input.text.toString()
                
                reciboActualizado?.let { recibo ->
                    PrinterFunctions.imprimirReimpRecibosTotal(
                        recibo,
                        requireActivity(),
                        1,
                        cantidadImpresiones
                    )
                    Toast.makeText(requireActivity(), "Reimprimir recibo", Toast.LENGTH_SHORT).show()
                }
            }
            setNegativeButton("Cancelar") { dialog, _ ->
                dialog.cancel()
            }
        }
        
        val alertDialog = alertDialogBuilder.create()
        alertDialog.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
        alertDialog.show()
    }

    private fun renderHtmlPreview() {
        try {
            val realm = Realm.getDefaultInstance()
            try {
                val preview = buildPreviewContent(realm)
                binding?.htmlTextReciboNuevo?.setHtmlFromString(preview, LocalImageGetter())
            } finally {
                realm.close()
            }
        } catch (e: Exception) {
            Log.e("ReciboResumen", "Error al generar vista previa", e)
            val errorMessage = "<center><h2>Error al cargar la vista previa</h2></center>"
            binding?.htmlTextReciboNuevo?.setHtmlFromString(errorMessage, LocalImageGetter())
        }
    }
    
    private fun buildPreviewContent(realm: Realm): String {
        if (reciboActualizado == null) {
            return "<center><h2>Seleccione un recibo para ver el detalle</h2></center>"
        }
        
        val customerId = reciboActualizado?.customer_id
        val cliente = realm.where(Clientes::class.java)
            .equalTo("id", customerId)
            .findFirst()
            
        val nombreCliente = cliente?.fantasyName ?: "Cliente no encontrado"
        
        var preview = "<h5>Recibos</h5>"
        preview += "<a><b>A nombre de:</b> $nombreCliente</a><br><br>"
        preview += "<a>------------------------------------------------<a><br>"
        preview += getReciboDetails(customerId ?: "", realm)
        
        return preview
    }

    private fun getReciboDetails(idCliente: String, realm: Realm): String {
        val reference = reciboActualizado?.reference ?: return "No hay datos de referencia"
        
        val receiptsResult = realm.where(receipts::class.java)
            .equalTo("customer_id", idCliente)
            .equalTo("reference", reference)
            .findAll()
            
        if (receiptsResult.isEmpty()) {
            return "No hay recibos emitidos para este cliente"
        }
        
        val receipt = receiptsResult.first()
        val recibo = realm.where(recibos::class.java)
            .equalTo("numeration", receipt.numeration)
            .findFirst()
            
        // Formatear datos del recibo
        val numeroReferencia = receipt.reference ?: ""
        val numeracion = receipt.numeration ?: ""
        val pagado = String.format("%,.2f", receipt.montoCanceladoPorFactura)
        val total = String.format("%,.2f", recibo?.total ?: 0.0)
        val restante = String.format("%,.2f", receipt.porPagarReceipts)
        val totalAbonos = String.format("%,.2f", receipt.balance)
        
        // Construir HTML con datos formateados
        return buildHtmlForRecibo(
            numeroReferencia,
            numeracion,
            pagado,
            total,
            totalAbonos,
            restante
        )
    }
    
    private fun buildHtmlForRecibo(
        referencia: String,
        numeracion: String,
        pagado: String,
        total: String,
        abonos: String,
        restante: String
    ): String {
        return "<a><b>" + padRight("# Referencia:", 20.0) + "</b></a>" + padRight(referencia, 20.0) + "<br>" +
                "<a><b>" + padRight("# Factura:", 30.0) + "</b></a>" + padRight(numeracion, 20.0) + "<br>" +
                "<a><b>" + padRight("Total Recibo:", 20.0) + "</b></a>" + padRight(pagado, 20.0) + "<br>" +
                "<a><b>" + padRight("Monto total:", 30.0) + "</b></a>" + padRight(total, 20.0) + "<br>" +
                "<a><b>" + padRight("Total en abonos:", 20.0) + "</b></a>" + padRight(abonos, 20.0) + "<br>" +
                "<a><b>" + padRight("Total restante:", 25.0) + "</b></a>" + padRight(restante, 20.0) + "<br>" +
                "<a>------------------------------------------------<a><br>"
    }
    
    private fun padRight(s: String, n: Double): String {
        return String.format("%-${n.toInt()}s", s)
    }

    override fun updateData() {
        val currentActivity = activity as? ReimprimirRecibosActivity
        if (currentActivity?.selecReciboTab == 1) {
            val reciboId = currentActivity.invoiceIdReimprimirRecibo
            
            val realm = Realm.getDefaultInstance()
            try {
                reciboActualizado = realm.where(receipts::class.java)
                    .equalTo("reference", reciboId)
                    .findFirst()?.let { realm.copyFromRealm(it) }
                
                Log.d("ReciboResumen", "ID Cliente: ${reciboActualizado?.customer_id}")
                Log.d("ReciboResumen", "ID Recibo: $reciboId")
                
                renderHtmlPreview()
            } finally {
                realm.close()
            }
        }
    }
    
    companion object {
        fun newInstance(): ReimprimirReciboResumenFragment = ReimprimirReciboResumenFragment()
    }
}


